optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/rules/AbstractGroupByMutator.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.core.impl.score.stream.drools.common.rules;
+
+import static java.util.Arrays.copyOfRange;
+import static java.util.Collections.emptyMap;
+import static java.util.Collections.singletonList;
+import static org.drools.model.DSL.accFunction;
+import static org.drools.model.PatternDSL.alphaIndexedBy;
+import static org.drools.model.PatternDSL.pattern;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Supplier;
+
+import org.drools.model.DSL;
+import org.drools.model.Index;
+import org.drools.model.PatternDSL;
+import org.drools.model.PatternDSL.PatternDef;
+import org.drools.model.Variable;
+import org.drools.model.view.ViewItem;
+import org.optaplanner.core.api.function.TriFunction;
+import org.optaplanner.core.impl.score.stream.drools.common.BiTuple;
+import org.optaplanner.core.impl.score.stream.drools.common.DroolsAbstractAccumulateFunction;
+import org.optaplanner.core.impl.score.stream.drools.common.DroolsAbstractGroupByAccumulator;
+import org.optaplanner.core.impl.score.stream.drools.common.FactTuple;
+import org.optaplanner.core.impl.score.stream.drools.common.QuadTuple;
+import org.optaplanner.core.impl.score.stream.drools.common.TriTuple;
+
+abstract class AbstractGroupByMutator implements Mutator {
+
+    protected abstract <InTuple> PatternDef bindTupleVariableOnFirstGrouping(AbstractRuleAssembler ruleAssembler,
+            PatternDef pattern, Variable<InTuple> tupleVariable);
+
+    protected ViewItem<?> getInnerAccumulatePattern(AbstractRuleAssembler ruleAssembler) {
+        List<ViewItem> allPatterns = new ArrayList<>();
+        for (int i = 0; i < ruleAssembler.getPrimaryPatterns().size(); i++) {
+            allPatterns.add(ruleAssembler.getPrimaryPatterns().get(i));
+            allPatterns.addAll(ruleAssembler.getDependentExpressionMap().getOrDefault(i, Collections.emptyList()));
+        }
+        ViewItem[] items = allPatterns.toArray(new ViewItem[0]);
+        return PatternDSL.and(items[0], copyOfRange(items, 1, items.length));
```

Comment:
Seems a bit fishy: A list (naming btw) that is converted into an array that is then copied partially